### PR TITLE
stream/field name validation/filtering

### LINKED ISSUE #954: Discovery should fail if the source contains illegal stream or field names
## Tell us about the problem you're trying to solve
Currently syncs transparently fail if the destination is unable to handle a stream or field with an illegal name. It would be better to circumvent this issue early in the proces.

### LINKED ISSUE #972: Sync worker should filter out streams and fields with illegal names
## Tell us about the problem you're trying to solve
As specified in the [Airbyte Protocol](https://docs.airbyte.io/architecture/airbyte-specification#discover), Airbyte only allows alphanumeric or underscore characters to be used in stream and field names. Thus, the sync worker should filter out streams with illegal stream/field names to prevent issues downstream.

### PR DESCRIPTION
**Pull request by jrhizor**
Resolves https://github.com/airbytehq/airbyte/issues/954
Resolves https://github.com/airbytehq/airbyte/issues/972

Start at `CatalogHelpers.java`.

Might need to address any discovery failures for existing API sources if they fail on integration tests here.

### CODE REVIEW

**Review by cgardens — APPROVED** [review]
This looks great. Can you add a page in our docs that explains the rules: 1. describes the allowed pattern 2. that we fail in discover 3. that we filter in sync? you can do as separate PR if you prefer, but i don't want us to forget to document this!

File: airbyte-workers/src/main/java/io/airbyte/workers/DefaultDiscoverCatalogWorker.java [hunk 1]
```diff
@@ -87,7 +90,21 @@ public DefaultDiscoverCatalogWorker(final IntegrationLauncher integrationLaunche
...
-          LOGGER.error("integration failed to output a catalog struct.");
+          LOGGER.error("Integration failed to output a catalog struct.");
+          return new OutputAndStatus<>(JobStatus.FAILED);
+        }
+
+        List<String> invalidStreamNames = CatalogHelpers.getInvalidStreamNames(catalog.get());
+
+        if (!invalidStreamNames.isEmpty()) {
+          invalidStreamNames.forEach(streamName -> LOGGER.error("Cannot sync invalid stream name: " + streamName));
+          return new OutputAndStatus<>(JobStatus.FAILED);
```
Comment (sherifnada) [inline airbyte-workers/src/main/java/io/airbyte/workers/DefaultDiscoverCatalogWorker.java:101]: can we also return a message describing why discovery failed so we can show this on the UI?

**Review by sherifnada — APPROVED** [review]

File: airbyte-protocol/models/src/test/resources/valid_schema.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (michel-tricot) [inline airbyte-protocol/models/src/test/resources/valid_schema.json:1]: We should have a test with Unicode

**Review by michel-tricot — APPROVED** [review]

File: airbyte-workers/src/main/java/io/airbyte/workers/DefaultDiscoverCatalogWorker.java (hunk 1 above)
Comment (jrhizor) [reply airbyte-workers/src/main/java/io/airbyte/workers/DefaultDiscoverCatalogWorker.java:101]: `OutputAndStatus` doesn't allow for error messages unfortunately. Adding messages outside of logs seems a bit out of scope for this PR. https://github.com/airbytehq/airbyte/issues/1011

File: airbyte-protocol/models/src/test/resources/valid_schema.json (hunk 2 above)
Comment (jrhizor) [reply airbyte-protocol/models/src/test/resources/valid_schema.json:1]: added

File: airbyte-workers/src/main/java/io/airbyte/workers/DefaultDiscoverCatalogWorker.java (hunk 1 above)
Comment (sherifnada) [reply airbyte-workers/src/main/java/io/airbyte/workers/DefaultDiscoverCatalogWorker.java:101]: I understand that, but I'm just worried that in its current state, there is no way for the user to know what's wrong. It'll just look like airbyte is broken if e.g: they sync Gsheets with a space in the name, and all they'll see on the UI is "an error happened"

**Pull request merged**